cpp/src/arrow/filesystem/s3fs.h

```diff
@@ -91,6 +94,12 @@ struct ARROW_EXPORT S3Options {
   /// Whether OutputStream writes will be issued in the background, without blocking.
   bool background_writes = true;
 
+  /// True when using SimpleAWSCredentialsProvider, else false.
+  bool creds_provided = false;
+
+  /// True when using web identity token to assume role
+  bool use_web_identity = false;
```

Comment:
Aren't `anonymous`, `creds_provided` and `use_web_identity` all exclusive? If so, how about replacing them with an enum, e.g.:
```c++
struct S3Options {
  enum CredentialsKind {
    /// Anonymous access (no credentials used)
    Anonymous,
    /// Use default AWS credentials, configured through environment variables
    Default,
    /// Use explicitly-provided username and password
    Explicit,
    /// Use web identity token to assume role, configured through environment variables
    WebIdentity
  };

  CredentialsKind credentials_kind = Default;
```

Reply:
Yes, that'a good idea. Thanks.

Reply:
@pitrou Done. Apologies for the delay.